lib/swoosh/adapters/smtp/helpers.ex

```diff
@@ -67,50 +67,42 @@ if Code.ensure_loaded?(:mimemail) do
     end
 
     defp prepare_parts(headers, %{
-      attachments: attachments,
-      html_body: nil,
-      text_body: text_body
-    }) when length(attachments) > 0 do
-      parts = Enum.map(attachments, &prepare_attachment(&1))
-      parts = [prepare_part(:plain, text_body) | parts]
-
-      {"multipart", "mixed", headers, parts}
-    end
-    defp prepare_parts(headers, %{
-      attachments: attachments,
+      attachments: [],
       html_body: html_body,
-      text_body: nil
-    }) when length(attachments) > 0 do
-      parts = Enum.map(attachments, &prepare_attachment(&1))
-      parts = [prepare_part(:html, html_body) | parts]
-
-      {"multipart", "mixed", headers, parts}
+      text_body: text_body
+    }) do
+      case {text_body, html_body} do
+        {text_body, nil} ->
+          headers = [{"Content-Type", "text/plain; charset=\"utf-8\""} | headers]
+          {"text", "plain", headers, text_body}
+        {nil, html_body} ->
+          headers = [{"Content-Type", "text/html; charset=\"utf-8\""} | headers]
+          {"text", "html", headers, html_body}
+        {text_body, html_body} ->
+          parts = [prepare_part(:plain, text_body), prepare_part(:html, html_body)]
+          {"multipart", "alternative", headers, parts}
+      end
     end
     defp prepare_parts(headers, %{
       attachments: attachments,
       html_body: html_body,
       text_body: text_body
-    }) when length(attachments) > 0 do
-      parts = Enum.map(attachments, &prepare_attachment(&1))
-      text_part = prepare_part(:plain, text_body)
-      html_part = prepare_part(:html, html_body)
-      alt_part = {"multipart", "alternative", [], [], [text_part, html_part]}
-
-      {"multipart", "mixed", headers, [alt_part | parts]}
-    end
-    defp prepare_parts(headers, %{html_body: nil, text_body: text_body}) do
-      headers = [{"Content-Type", "text/plain; charset=\"utf-8\""} | headers]
-      {"text", "plain", headers, text_body}
-    end
-    defp prepare_parts(headers, %{html_body: html_body, text_body: nil}) do
-      headers = [{"Content-Type", "text/html; charset=\"utf-8\""} | headers]
-      {"text", "html", headers, html_body}
-    end
-    defp prepare_parts(headers, %{html_body: html_body, text_body: text_body}) do
-      parts = [prepare_part(:plain, text_body), prepare_part(:html, html_body)]
-      {"multipart", "alternative", headers, parts}
+    }) do
+      content_part =
+        case {prepare_part(:plain, text_body), prepare_part(:html, html_body)} do
+          {text_part, nil} ->
+            text_part
+          {nil, html_part} ->
+            html_part
+          {text_part, html_part} ->
+            {"multipart", "alternative", [], [], [text_part, html_part]}
+        end
+      attachment_parts = Enum.map(attachments, &prepare_attachment(&1))
+
+      {"multipart", "mixed", headers, [content_part | attachment_parts]}
     end
 
+    defp prepare_part(_subtype, nil), do: nil
```

Comment:
maybe is worth checking also for the empty string? just in case of poor usage by the user...

Reply:
Might be completely wrong here but I think we're only checking `nil` so that user can still have empty parts (allowed by the spec).

Reply:
Hmm, I'm a little skeptical about changing behavior in this PR. Though not probable, someone could be using it as a feature 😉

Reply:
sure, if the standard allows that, I'm fine with it :)